haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs

```diff
@@ -841,6 +841,12 @@ Notification:
 
 type TelemetryNotification = NotificationMessage ServerMethod A.Value
 
+type CustomNotification = NotificationMessage Text A.Value
```

Comment:
Shouldn't this be
```suggestion
type CustomNotification = NotificationMessage ClientMethod A.Value
```

Reply:
That would require adding a new constructor to `ClientMethod` that handles this. I don’t think we gain anything from this. This really can be an arbitrary Text. The reason for why we need to add the corresponding constructor to `ServerMethod` is that we need to be able to match on it in `handlerMap`. This does not apply here.

Reply:
Still, it might be worth it for the sake of consistency, everything else is of the form `NotificationMessage ClientMethod` or `NotificationMessage ServerMethod`, and I believe there are functions for sending/receiving messages that constrain this type parameter to `ClientMethod` or `ServerMethod` so that you don't accidentally send/receive the wrong kind of message.

Reply:
> Isn't `CustomClientMessage` the required `ClientMethod` constructor?

Oh right, I mixed things up. I’m happy to change it to `ClientMethod` if you and @alanz prefer ithat but personally I think that’s more confusing than helpful. It suggests that it can be any of the constructors of `ClientMethod` when really we only want to allow for `CustomClientMethod`.

Reply:
> It suggests that it can be any of the constructors of ClientMethod when really we only want to allow for CustomClientMethod.

Yes, but it is like that for all of the other types too. This was a compromise as opposed to a more heavyweight system with singletons.

This way, we get one extra bit of information at the type level(whether something is a ClientMethod or a ServerMethod).

Reply:
Alright, I’ve switched to using `ClientMethod` and `ServerMethod`.